src/opm/input/eclipse/Schedule/UDQ/UDQEnums.cpp

```diff
@@ -435,131 +452,151 @@ UDAKeyword keyword(UDAControl control)
-        {UDAControl::GCONINJE_SURFACE_MAX_RATE,      300017}, // Surface injection rate
-        {UDAControl::GCONINJE_RESV_MAX_RATE,         400017}, // Reservoir volume injection rate
-        {UDAControl::GCONINJE_TARGET_REINJ_FRACTION, 500017}, // Reinjection fraction
-        {UDAControl::GCONINJE_TARGET_VOID_FRACTION,  600017}, // Voidage replacement fraction
+        {UDAControl::WELTARG_ORAT,        16},
+        {UDAControl::WELTARG_WRAT,   100'016},
+        {UDAControl::WELTARG_GRAT,   200'016},
+        {UDAControl::WELTARG_LRAT,   300'016},
+        {UDAControl::WELTARG_RESV,   400'016},
+        {UDAControl::WELTARG_BHP,    500'016},
+        {UDAControl::WELTARG_THP,    600'016},
+        {UDAControl::WELTARG_LIFT, 1'000'016},
     };
 
     return lookup_control_map_value(c2uda, control);
 }
 
-bool group_control(UDAControl control) {
-    if (control == UDAControl::GCONPROD_OIL_TARGET)
-        return true;
-
-    if (control == UDAControl::GCONPROD_WATER_TARGET)
-        return true;
-
-    if (control == UDAControl::GCONPROD_GAS_TARGET)
-        return true;
-
-    if (control == UDAControl::GCONPROD_LIQUID_TARGET)
-        return true;
-
-    if (control == UDAControl::GCONINJE_SURFACE_MAX_RATE)
-        return true;
-
-    if (control == UDAControl::GCONINJE_RESV_MAX_RATE)
-        return true;
-
-    if (control == UDAControl::GCONINJE_TARGET_REINJ_FRACTION)
-        return true;
-
-    if (control == UDAControl::GCONINJE_TARGET_VOID_FRACTION)
-        return true;
-
-    return false;
+bool group_control(const UDAControl control)
+{
+    try {
+        const auto kw = keyword(control);
+        return (kw == UDAKeyword::GCONPROD)
+            || (kw == UDAKeyword::GCONINJE);
+    }
+    catch (const std::logic_error&) {
+        return false;
+    }
 }
 
-bool well_control(UDAControl control) {
-    return !group_control(control);
+bool well_control(const UDAControl control)
+{
+    try {
+        const auto kw = keyword(control);
+        return (kw == UDAKeyword::WCONPROD)
+            || (kw == UDAKeyword::WCONINJE)
+            || (kw == UDAKeyword::WELTARG);
+    }
+    catch (const std::logic_error&) {
+        return false;
+    }
 }
 
-bool injection_control(UDAControl control) {
-    switch (control) {
-    case UDAControl::WCONINJE_RATE:
-    case UDAControl::WCONINJE_RESV:
-    case UDAControl::WCONINJE_BHP:
-    case UDAControl::WCONINJE_THP:
-    case UDAControl::GCONINJE_SURFACE_MAX_RATE:
-    case UDAControl::GCONINJE_RESV_MAX_RATE:
-    case UDAControl::GCONINJE_TARGET_REINJ_FRACTION:
-    case UDAControl::GCONINJE_TARGET_VOID_FRACTION:
-        return true;
-    default:
+bool is_weltarg(const UDAControl control)
+{
+    try {
+        return keyword(control) == UDAKeyword::WELTARG;
+    }
+    catch (const std::logic_error&) {
         return false;
     }
 }
 
-bool production_control(UDAControl control) {
-    return !injection_control(control);
+bool injection_control(const UDAControl control)
```

Comment:
It is not good that injection_control() can return false when the control is a WELTARG control on an injector.

Suggest splitting into is_injection_well_control(const UDAControl, const Well&) and is_injection_group_control(const UDAControl). Similar for production.

Reply:
> It is not good that injection_control() can return false when the control is a WELTARG control on an injector.

Agreed.

> Suggest splitting into `is_injection_well_control(const UDAControl, const Well&)` and `is_injection_group_control(const UDAControl)`. Similar for production.

I ended up creating a slight variation of these,
```C++
bool is_well_injection_control(const UDAControl, const bool isInjector);
bool is_well_production_control(const UDAControl, const bool isProducer);
bool is_group_injection_control(const UDAControl);
bool is_group_production_control(const UDAControl);
```
mostly to avoid having to make the `UDQEnums.hpp` and `UDQEnums.cpp` files aware of `Opm::Well`.  These enum files are also used in the code generator for the input parser and I'd prefer to avoid cyclic dependencies.